platform/platform-crypto-provider/src/main/java/org/codice/ddf/platform/crypto/provider/MerlinExtension.java

```diff
@@ -0,0 +1,340 @@
+                    X509Certificate[] x509certs = new X509Certificate[foundCertChain.length + 1];
+                    x509certs[0] = certs[0];
+                    System.arraycopy(foundCertChain, 0, x509certs, 1, foundCertChain.length);
+
+                    List<X509Certificate> certList = Arrays.asList(certs);
+                    CertPath path = getCertificateFactory().generateCertPath(certList);
+
+                    try {
+                        validator.validate(path, param);
+                        // We have a valid cert path at this point so break
+                        validatorException = null;
+                        break;
+                    } catch (java.security.cert.CertPathValidatorException e) {
+                        validatorException = e;
+                    }
+                }
+
+                if (validatorException != null) {
+                    throw validatorException;
+                }
+            } else {
+                List<X509Certificate> certList = Arrays.asList(certs);
+                CertPath path = getCertificateFactory().generateCertPath(certList);
+
+                validator.validate(path, param);
+            }
+        } catch (NoSuchProviderException | NoSuchAlgorithmException
+                | CertificateException | InvalidAlgorithmParameterException
+                | java.security.cert.CertPathValidatorException
+                | KeyStoreException e) {
+            throw new WSSecurityException(
+                    WSSecurityException.ErrorCode.FAILURE, e, "certpath"
+            );
+        }
+
+        // Finally check Cert Constraints
+        if (!matchesSubjectDnPattern(certs[0], subjectCertConstraints)) {
+            throw new WSSecurityException(WSSecurityException.ErrorCode.FAILED_AUTHENTICATION);
+        }
+    }
+
+    private void buildTrustAnchors(Set<TrustAnchor> trustAnchorSet,
+            Enumeration<String> truststoreAliases) throws KeyStoreException {
+        while (truststoreAliases.hasMoreElements()) {
+            String alias = truststoreAliases.nextElement();
+            X509Certificate cert = (X509Certificate) truststore.getCertificate(alias);
+            if (cert != null) {
+                byte[] nameConstraints = new byte[0];
+                try {
+                    nameConstraints = extractNameConstraints(cert);
+                } catch (IOException e) {
+                    LOG.debug("Unable to decode name constraints.", e);
+                }
+                TrustAnchor trustAnchor = new TrustAnchor(cert, nameConstraints);
+                trustAnchorSet.add(trustAnchor);
+            }
+        }
+    }
+
+    private static byte[] extractNameConstraints(X509Certificate x509Certificate)
+            throws IOException {
+        byte[] bytes = x509Certificate.getExtensionValue(NAME_CONSTRAINTS_OID);
+        if (bytes == null || bytes.length <= 0) {
+            return null;
```

Comment:
❓ Here you'll return null, but in the case of an IOException up on line 246 you'll log and pass a zero-byte array to the `TrustAnchor`. I'm not sure which is better, but I think they should probably be the same.